Do not load chunks from loading queue if they are not needed anymore

wonder-sk opened the pull request:
When user moves camera in 3D view, we may end up adding quite a few chunks to the loading queue. Often by the time they get loaded (especially in case of remote sources), the camera is already looking at something different and it was a waste of network + CPU + GPU memory to load those chunks. This PR tries to avoid that by going through the loading queue, and it removes any nodes that fail the frustum culling tests (i.e. camera is not looking at those nodes anymore).

~~(work in progress - there seem to be some minor bugs - when 3D renderer is reloaded, it may get "stuck" and do not load chunks and user needs to zoom out & zoom in to make things work again)~~

@NEDJIMAbelgacem

wonder-sk committed:
Do load chunks from loading queue if they are not needed anymore (wip)

nyalldawson commented on src/3d/chunks/qgschunkedentity_p.cpp:
```diff
@@ -176,7 +176,43 @@ void QgsChunkedEntity::update( const SceneState &state )
+      n->unloadChunk();
+    }
+  }
+  qDebug() << "removed useless tiles in loading queue" << uselessTiles;
```
Can we ifdef these out ?

wonder-sk committed:
Fix a bug with loading queue pruning and refactor code a bit

The update scene function was getting the logic a bit complicated
with additive/replacement strategies mixed together, not it should
be hopefully easier to follow.

wonder-sk committed:
Remove refactored code after rebase

wonder-sk commented:
Ready for review now...

uclaros commented on src/3d/chunks/qgschunkedentity_p.cpp:
```diff
@@ -339,32 +387,64 @@ void QgsChunkedEntity::update( QgsChunkNode *root, const SceneState &state )
+      // This chunk does not have acceptable error (it does not provide enough detail)
+      // so we'll try to use its children. The exact logic depends on whether the entity
+      // has additive strategy. With additive strategy, child nodes should be rendered
+      // it addition to the parent nodes (rather than child nodes replacing parent entirely)
```
```suggestion
      // in addition to the parent nodes (rather than child nodes replacing parent entirely)
```

uclaros reviewed (approved):
Nicely done!
I love the extensive comments describing the algorithm. It's really nice seeing the programmer's intentions when revisiting code in the future!

just a super minor typo in the comments

wonder-sk committed:
Update src/3d/chunks/qgschunkedentity_p.cpp

Co-authored-by: Stefanos Natsis <uclaros@gmail.com>
